flink-runtime-web/src/main/java/org/apache/flink/runtime/webmonitor/metrics/MetricFetcher.java

```diff
@@ -0,0 +1,185 @@
+				fetchMetrics();
+			}
+		}
+	}
+
+	private void fetchMetrics() {
+		try {
+			Option<scala.Tuple2<ActorGateway, Integer>> jobManagerGatewayAndWebPort = retriever.getJobManagerGatewayAndWebPort();
+			if (jobManagerGatewayAndWebPort.isDefined()) {
+				ActorGateway jobManager = jobManagerGatewayAndWebPort.get()._1();
+
+				/**
+				 * Remove all metrics that belong to a job that is not running and no longer archived.
+				 */
+				jobManager.ask(new RequestJobDetails(true, true), timeout)
+					.onSuccess(new OnSuccess<Object>() {
+						@Override
+						public void onSuccess(Object result) throws Throwable {
+							MultipleJobsDetails details = (MultipleJobsDetails) result;
+							ArrayList<String> toRetain = new ArrayList<>();
+							for (JobDetails job : details.getRunningJobs()) {
+								toRetain.add(job.getJobId().toString());
+							}
+							for (JobDetails job : details.getFinishedJobs()) {
+								toRetain.add(job.getJobId().toString());
+							}
+							synchronized (metrics) {
+								metrics.jobs.keySet().retainAll(toRetain);
+							}
+						}
+					}, ctx);
+
+				/**
+				 * Requests the metric dump from the job manager.
+				 */
+				jobManager.ask(Messages.getRequestMetrics(), timeout)
+					.onSuccess(new OnSuccess<Object>() {
+						@Override
+						public void onSuccess(Object result) throws Throwable {
+							addMetrics(result);
+						}
+					}, ctx);
+
+				/**
+				 * We first request the list of all registered task managers from the job manager, and then
+				 * request the respective metric dump from each task manager.
+				 *
+				 * All stored metrics that do not belong to a registered task manager will be removed.
+				 */
+				jobManager.ask(JobManagerMessages.getRequestRegisteredTaskManagers(), timeout)
+					.onSuccess(new OnSuccess<Object>() {
+						@Override
+						public void onSuccess(Object result) throws Throwable {
+							Iterable<Instance> taskManagers = ((JobManagerMessages.RegisteredTaskManagers) result).asJavaIterable();
+							List<String> activeTaskManagers = new ArrayList<>();
+							for (Instance taskManager : taskManagers) {
+								activeTaskManagers.add(taskManager.getId().toString());
+								taskManager.getActorGateway().ask(Messages.getRequestMetrics(), timeout)
+									.onSuccess(new OnSuccess<Object>() {
+										@Override
+										public void onSuccess(Object result) throws Throwable {
+											addMetrics(result);
+										}
+									}, ctx);
+							}
+							synchronized (metrics) { // remove all metrics belonging to unregistered task managers
+								metrics.taskManagers.keySet().retainAll(activeTaskManagers);
+							}
+						}
+					}, ctx);
+			}
+		} catch (Exception e) {
+			LOG.warn("Exception while fetching metrics.", e);
+		}
+	}
+
+	private void addMetrics(Object result) {
+		Object[] data = (Object[]) result;
+		for (int x = 0; x < data.length; ) {
+			synchronized (metrics) {
+				switch ((byte) data[x++]) {
+					case 0:
+					case 1:
```

Comment:
Could we introduce some enums or so for these values. Maybe we could use the byte value as the ordinal of the corresponding enum value.